File: 프로그래머스/unrated/150370. 개인정보 수집 유효기간/개인정보 수집 유효기간.js
const solution = (today, terms, privacies) => {
    const todayDate = today.split(".").map(Number)
    
    // term 정리
    const termDict = {}
    for(const term of terms){
        const tmp = term.split(" ")
        termDict[tmp[0]] = Number(tmp[1])
    }
    
    return privacies.map((privacy, i) => {
        const [privacyDate, termType] = privacy.split(" ")
        const curDate = privacyDate.split(".").map(Number)
        
        // 유효기간 더하기
        
        curDate[1] = curDate[1] + termDict[termType]
        if (curDate[1] > 12){
            curDate[0]++
            curDate[1] -= 12
        }
        
        // 기간 비교
        const todaySum = todayDate[0] * 12 * 28 + todayDate[1] * 28 + todayDate[2]
        const curSum = curDate[0] * 12 * 28 + curDate[1] * 28 + curDate[2]
        if (todaySum >= curSum) return i+1
        
    }).filter(x => x)
    
}

File: 프로그래머스/unrated/150370. 개인정보 수집 유효기간/개인정보 수집 유효기간.py
def createTermDict(terms):
    tmpDict = dict()
    for term in terms:
        [term_type, month] = term.split(" ")
        tmpDict[term_type] = int(month)
    return tmpDict


def solution(today, terms, privacies):
    ans = []
    [cur_year, cur_month, cur_day] = map(int, today.split("."))
    termDict = createTermDict(terms)  # term 딕셔너리 만들기
    
    for index, privacy in enumerate(privacies):
        [date, term_type] = privacy.split(" ")
        [year, month, day] = map(int, date.split("."))
        month += termDict[term_type]
        day -= 1
        while month > 12:
            year += 1
            month -= 12
        
        date_sum = year * 12 * 28 + month * 28 + day
        cur_date_sum = cur_year * 12 * 28 + cur_month * 28 + cur_day
        if date_sum < cur_date_sum:
            ans.append(index + 1)
    
    return ans